Using a FQN for a class defined in another file compiles but then it cannot be resolved in runtime

Sample

Main file having the issue

**a.wpgm**

``` xtend
program zuper {
        val pepona = new b.Golondrina()
}
```

Another file defines the class: 

**b.wlk**

``` xtend
class Golondrina  {
    method volar(kms) {
        console.println("flying...")
    }
}
```

This compiles fine (passes the validations), and also you can navigate the constructor call to the class Golondrina, but if you run the program you'll get something like

```
[ERROR] /Users/jfernandes/dev/data/repo/wollok-dev/wollok-tests/src/wollok/classes/super-program.wpgm:7:89 ERROR Couldn't resolve reference to WClass 'b.Golondrina'.
```

--- a/org.uqbar.project.wollok.tests/src/org/uqbar/project/wollok/tests/interpreter/imports/ImportsTest.xtend
+++ b/org.uqbar.project.wollok.tests/src/org/uqbar/project/wollok/tests/interpreter/imports/ImportsTest.xtend
@@ -382,12 +382,14 @@ class ImportsTest extends AbstractWollokInterpreterTestCase {
 				}
 			}
 		'''
-		].interpretPropagatingErrors
+		].interpretAsFilesPropagatingErrors
 	}
 	
 	@Test
 	def void testRefereceByFQNWithoutImport() {
-		#['b' -> '''
+		
+		#[
+		'b' -> '''
 			class Golondrina  {
 			    method volar(kms) {
 			        console.println("flying...")
@@ -399,7 +401,7 @@ class ImportsTest extends AbstractWollokInterpreterTestCase {
 			        const pepona = new b.Golondrina()
 			}
 		'''
-		].interpretPropagatingErrors
+		].interpretAsFilesPropagatingErrors
 	}
 	
 }

--- a/org.uqbar.project.wollok/src/org/uqbar/project/wollok/model/WollokModelExtensions.xtend
+++ b/org.uqbar.project.wollok/src/org/uqbar/project/wollok/model/WollokModelExtensions.xtend
@@ -65,6 +65,8 @@ import org.eclipse.emf.ecore.resource.Resource
 import static org.uqbar.project.wollok.scoping.root.WollokRootLocator.*
 
 import static extension org.uqbar.project.wollok.model.WMethodContainerExtensions.*
+import org.eclipse.xtext.nodemodel.util.NodeModelUtils
+import org.uqbar.project.wollok.wollokDsl.impl.ImportImpl
 
 /**
  * Extension methods to Wollok semantic model.
@@ -459,6 +461,16 @@ class WollokModelExtensions {
 	def static Iterable<Import> allImports(EObject e){ 
 		val locals = e.eContents.filter(Import)
 		if(e.eContainer !== null) e.eContainer.allImports() + locals else locals
+	}
+	
+	def static Iterable<String> allFQNImports(EObject e){
+		val constructors = e.eAllContents.filter(WConstructorCall).toSet
+		constructors
+			.map[NodeModelUtils.findNodesForFeature(it,WollokDslPackage.Literals.WCONSTRUCTOR_CALL__CLASS_REF)]
+			.flatten
+			.map[NodeModelUtils.getTokenText(it)]
+			.filter[it.contains(".")].toSet
+		#{}
 	}
 	
 	// *******************************

--- a/org.uqbar.project.wollok/src/org/uqbar/project/wollok/scoping/WollokGlobalScopeProvider.xtend
+++ b/org.uqbar.project.wollok/src/org/uqbar/project/wollok/scoping/WollokGlobalScopeProvider.xtend
@@ -59,19 +59,19 @@ class WollokGlobalScopeProvider extends DefaultGlobalScopeProvider {
 	 */
 	def importedObjects(Resource context) {
 		val rootObject = context.contents.get(0)
-		val imports = rootObject.getAllContentsOfType(Import)
+		val imports = rootObject.allImports.map[importedNamespace] + rootObject.allFQNImports
 		cache.get(context.URI, imports, [doImportedObjects(context, imports)])
 		// doImportedObjects(context,imports)
 	}
 
-	def doImportedObjects(Resource context, List<Import> imports) {
+	def doImportedObjects(Resource context, Iterable<String> imports) {
 		val objectsFromManifests = libraryLoader.load(context)
 
 		objectsFromLocalImport(context, imports, objectsFromManifests)
 	}
 	
-	def objectsFromLocalImport(Resource context, List<Import> importsEntry, Iterable<IEObjectDescription> objectsFromManifests){
-		val imports = (importsEntry.map[ #[importedNamespace]  + localScopeProvider.allRelativeImports(importedNamespace, context.implicitPackage) ].flatten).toSet
+	def objectsFromLocalImport(Resource context, Iterable<String> importsEntry, Iterable<IEObjectDescription> objectsFromManifests){
+		val imports = (importsEntry.map[ #[it]  + localScopeProvider.allRelativeImports(it, context.implicitPackage) ].flatten).toSet
 		
 		val importedObjects = imports.filter[
 			it !== null && !objectsFromManifests.exists[o| o.matchesImport(it)]

--- a/org.uqbar.project.wollok/src/org/uqbar/project/wollok/scoping/cache/MapBasedWollokGlobalScopeCache.xtend
+++ b/org.uqbar.project.wollok/src/org/uqbar/project/wollok/scoping/cache/MapBasedWollokGlobalScopeCache.xtend
@@ -1,8 +1,6 @@
 package org.uqbar.project.wollok.scoping.cache
 
 import com.google.inject.Singleton
-import java.util.List
-import java.util.Set
 import org.apache.commons.collections.map.LRUMap
 import org.eclipse.emf.common.util.URI
 import org.eclipse.xtend.lib.annotations.Accessors
@@ -15,11 +13,11 @@ class MapBasedWollokGlobalScopeCache implements WollokGlobalScopeCache {
 	
 	val cache = new LRUMap(20)
 	
-	override get(URI uri, List<Import> imports, Function0<Iterable<IEObjectDescription>> ifAbsentBlock) {
+	override get(URI uri, Iterable<String> imports, Function0<Iterable<IEObjectDescription>> ifAbsentBlock) {
 		val uriString = uri.toString
 		var cacheContent = cache.get(uriString) as MapBasedCacheContent
 		
-		if(cacheContent == null || cacheContent.mismatch(imports)){
+		if(cacheContent === null || cacheContent.mismatch(imports)){
 			cacheContent = new MapBasedCacheContent(uri, imports, ifAbsentBlock.apply)
 			cache.put(uriString, cacheContent)
 		}
@@ -36,16 +34,16 @@ class MapBasedWollokGlobalScopeCache implements WollokGlobalScopeCache {
 @Accessors
 class MapBasedCacheContent {
 	val String uri
-	val Set<String> imports
+	val Iterable<String> imports
 	val Iterable<IEObjectDescription> result
 	
-	new(URI uri, List<Import> imports, Iterable<IEObjectDescription> result){
+	new(URI uri, Iterable<String> imports, Iterable<IEObjectDescription> result){
 		this.uri = uri.toString
-		this.imports = imports.map[importedNamespace].toSet
+		this.imports = imports.toSet
 		this.result = newHashSet(result)
 	}
 	
-	def mismatch(List<Import> imports){
-		this.imports != imports.map[importedNamespace].toSet
+	def mismatch(Iterable<String> imports){
+		this.imports != imports.toSet
 	}
 }

--- a/org.uqbar.project.wollok/src/org/uqbar/project/wollok/scoping/cache/WollokGlobalScopeCache.xtend
+++ b/org.uqbar.project.wollok/src/org/uqbar/project/wollok/scoping/cache/WollokGlobalScopeCache.xtend
@@ -1,12 +1,10 @@
 package org.uqbar.project.wollok.scoping.cache
 
-import java.util.List
 import org.eclipse.emf.common.util.URI
-import org.eclipse.xtext.xbase.lib.Functions.Function0
-import org.uqbar.project.wollok.wollokDsl.Import
 import org.eclipse.xtext.resource.IEObjectDescription
+import org.eclipse.xtext.xbase.lib.Functions.Function0
 
 interface WollokGlobalScopeCache {
-	def Iterable<IEObjectDescription> get(URI uri, List<Import> imports, Function0<Iterable<IEObjectDescription>> ifAbsentBlock)
+	def Iterable<IEObjectDescription> get(URI uri, Iterable<String> imports, Function0<Iterable<IEObjectDescription>> ifAbsentBlock)
 	def void clearCache()
 }

